Add request_id option to client

funkyboy opened the pull request.

SimonWoolf commented on lib/ably/rest/client.rb:
```diff
@@ -146,6 +150,7 @@ def initialize(options)
         @custom_host      = options.delete(:rest_host)
         @custom_port      = options.delete(:port)
         @custom_tls_port  = options.delete(:tls_port)
+        @add_request_ids  = options.delete(:add_request_ids) == false ? false : true
```
As an optional extra debug option, this should default to false (unless and until it's added to the spec). So just `@add_request_ids  = options.delete(:add_request_ids)` is fine.

(Also, `cond ? false : true` is a pointless pattern, it's equivalent to `!cond`)

SimonWoolf commented on lib/ably/rest/client.rb:
```diff
@@ -434,6 +439,16 @@ def send_request(method, path, params, options)
+          path_with_params.query_values = {request_id: SecureRandom.hex(20)}
+          query = path_with_params.query
+          path = "#{path}?#{query}"
+        end
```
Why are you manually munging the path? This method takes a params hash. You can add it to the params hash.

mattheworiordan replied:
@SimonRichardson because this is Ruby and not Elixir.  So the `params` object would be changed by just calling this method, which is bad.

SimonWoolf replied:
> So the params object would be changed by just calling this method, which is bad.

@mattheworiordan fair enough, but imo the solution to that is to clone the hash (as Cesare has done in the latest version of this pr), not to start munging a param into the path (meaning faraday has to parse it out again to combine with the params -- assuming it even does that rather than give you a path with two `?`s). (for one thing the original approach assumes the path doesn't already have a querysting, so fails the 'what if two things did this' test)

SimonWoolf commented on spec/unit/rest/client_spec.rb:
```diff
@@ -50,4 +50,11 @@
+    it 'includes request_id in URL' do
+      expect(subject.add_request_ids).to eql(true)
+    end
+  end
```
If you're going to add a test, may as well be one that checks the result (i.e. uses webmock to check that a `request_id` actually gets added to the querystring for rest requests, and checks timeout exceptions to make sure that the request ID gets printed with those), rather than just checks that an instance variable got set correctly, no?

funkyboy commented:
@SimonWoolf re-ready for review.

mattheworiordan commented on lib/ably/exceptions.rb:
```diff
@@ -13,9 +13,9 @@ module Exceptions
+      attr_reader :status, :code, :request_id
 
-      def initialize(message, status = nil, code = nil, base_exception = nil, options = {})
+      def initialize(message, status = nil, code = nil, base_exception = nil, options = {}, request_id = nil)
```
I would have preferred for `request_id` to be an option given it's a flexible hash for this reason.

funkyboy replied:
Didn’t know the supposed usage of that hash, that’s why I didn’t touch it. Good to know that I can use it.

mattheworiordan commented on lib/ably/exceptions.rb:
```diff
@@ -71,8 +72,8 @@ class EncoderError < BaseAblyException; end
-      def initialize(message, status = nil, code = nil, base_exception = nil, options = {})
-        super message, status, code, base_exception, options
+      def initialize(message, status = nil, code = nil, base_exception = nil, options = {}, request_id = nil)
+        super message, status, code, base_exception, options, request_id
```
If we leave `request_id` as an option of options, then no change here is needed.

mattheworiordan commented on lib/ably/rest/client.rb:
```diff
@@ -434,6 +439,10 @@ def send_request(method, path, params, options)
         max_retry_duration = http_defaults.fetch(:max_retry_duration)
         requested_at       = Time.now
         retry_count        = 0
+        if @add_request_ids
```
It's bad practice to read an instance variable like this when it's available as a reader i.e. `if @add_request_idds` would always be false due to a type, whereas `if add_request_idds` would fail.

mattheworiordan commented on lib/ably/rest/client.rb:
```diff
@@ -456,12 +465,12 @@ def send_request(method, path, params, options)
           case error
             when Faraday::TimeoutError
-              raise Ably::Exceptions::ConnectionTimeout.new(error.message, nil, 80014, error)
+              raise Ably::Exceptions::ConnectionTimeout.new(error.message, nil, 80014, error, {}, request_id)
```
Based on my proposed change, it would be `raise Ably::Exceptions::ConnectionTimeout.new(error.message, nil, 80014, error, { request_id: request_id })`.
I think `raise Ably::Exceptions::ConnectionTimeout.new(error.message, nil, 80014, error, request_id: request_id)` may work too, although I recall 1.9.2 having some constraints with pseudo named args.

mattheworiordan commented on spec/unit/rest/client_spec.rb:
```diff
@@ -50,4 +50,11 @@
   end
+
+  context 'request_id generation' do
+    let(:client_options) { {key: 'appid.keyuid:keysecret', add_request_ids: true} }
```
Unusual spacing, in all other places `{` has a space after, and `}` a space before

funkyboy commented:
@SimonWoolf @mattheworiordan 

- added `request_id` for nonce errors
- now passing `request_id` in the `options` param
- added more tests for both nonce and timeout errors with `add_request_ids` true and false

Ready for another look.

mattheworiordan commented on spec/acceptance/rest/client_spec.rb:
```diff
@@ -971,6 +971,20 @@ def encode64(text)
+          begin
+            client2.auth.request_token(token_params)
+          rescue Ably::Exceptions::UnauthorizedRequest => err
+            expect(err.request_id).to_not eql(nil)
```
Do we have any test to ensure that when `add_request_ids` is false, that the `request_id` is `nil`?

mattheworiordan commented on lib/ably/rest/client.rb:
```diff
@@ -470,9 +470,9 @@ def send_request(method, path, params, options)
           case error
             when Faraday::TimeoutError
-              raise Ably::Exceptions::ConnectionTimeout.new(error.message, nil, 80014, error, {}, random_request_id)
+              raise Ably::Exceptions::ConnectionTimeout.new(error.message, nil, 80014, error, {request_id: random_request_id})
```
Spacing issues again, here and many times below

mattheworiordan replied:
I am not a big fan of this. `random_request_id` is defined in line 445, yet is within a conditional.  I realise Ruby allows this sort of thing, but it's messy and error prone.  Why are you defining `random_request_id` as a varaiable when you have access to the `params` variable in this scope which is available anyway?

mattheworiordan replied:
In fact, I am not sure I have seen a test that confirms that a) a request is sent to Ably with a `request_id`, b) it fails (intentionally), c) the test confirms that the exception raised has that SAME `request_id`, d) the Logger is showing the `request_id` in the output.  This is the crux of what we're doing here.  Given this code path, I fear our tests so far are just testing that a `request_id` is present, but not ensuring it is consistent throughout.

funkyboy replied:
This makes sense, but I am struggling a bit to inspect the url of a request, say `client.stats`. Any suggestion?

mattheworiordan commented on lib/ably/rest/middleware/exceptions.rb:
```diff
@@ -27,7 +27,7 @@ def on_complete(env)
 
             message = 'Unknown server error' if message.to_s.strip == ''
-            request_id = env.request.context[:request_id]
+            request_id = env.request.context[:request_id] if env.request.context
```
Spacing issues below.

mattheworiordan commented on lib/ably/rest/client.rb:
```diff
@@ -456,12 +465,12 @@ def send_request(method, path, params, options)
-              raise Ably::Exceptions::ConnectionError.new(error.message, nil, 80000, error)
+              raise Ably::Exceptions::ConnectionError.new(error.message, nil, 80000, error, {}, request_id)
             else
               raise error
```
I am also not keen on a var called `random_request_id`, that sounds more like a method.  It's no longer random once it's assigned, and thus it's consistent for the request.

mattheworiordan commented on spec/acceptance/rest/client_spec.rb:
```diff
@@ -963,16 +963,17 @@ def encode64(text)
-    end
-        context 'without request_id' do
+
+      context 'without request_id' do
```
This doesn't fix the issue?  @funkyboy what is happening here?  Can you explain what the problem is as this is the second attempt to fix indentation issues and it's failed.  What IDE are you using?

![capto_capture 2018-04-18_06-24-58_pm](https://user-images.githubusercontent.com/43789/38947880-f9ced374-4335-11e8-9549-d03bdba4613d.png)

funkyboy replied:
I am using VS code but clearly there’s something wrong going on. Will open an issue to add rubocop, so we don’t waste time on this.

funkyboy commented:
- [ ] Fix indentation (for real)
- [ ] Add test to verify that the `request_id` is the same at request and response time, and that it's included in the error and the logger
- [ ] Add test to verify that, when a request fails, it is retried to a different endpoint with the same `request_id`. See https://github.com/ably/docs/issues/359

mattheworiordan commented on lib/ably/rest/client.rb:
```diff
@@ -434,11 +439,22 @@ def send_request(method, path, params, options)
         retry_count        = 0
+        request_id         = nil
+        if add_request_ids
+          params = {} if params.nil?
```
This is a tiny comment, but it seems unnecesary to dup an empty hash, so perhaps instead just:

```ruby
params = if params.nil?
  {}
else
  params.dup
end
```

mattheworiordan commented on lib/ably/rest/client.rb:
```diff
@@ -456,12 +472,11 @@ def send_request(method, path, params, options)
+              raise Ably::Exceptions::ConnectionTimeout.new(error.message, nil, 80014, error, { request_id: request_id })
             when Faraday::ClientError
-              raise Ably::Exceptions::ConnectionError.new(error.message, nil, 80000, error)
+              raise Ably::Exceptions::ConnectionError.new(error.message, nil, 80000, error, { request_id: request_id })
```
May be nice for future us to add a note about the `raise error` not explicitly passing the `request_id`, but in the Middleware it's available through the requext context. WDYT?
